### PR TITLE
Add ECSClient module

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context
Add an `ECSClient` module that can run and monitor ECS tasks.

Integrating `webutils.aws.ecs.ECSClient` with the Flask application is **out of scope** for this PR. There will be a separate effort for workflow orchestration. That said, the orchestrator (TBD) is likely to use the client like so: 

```python
ecs_client = ECSClient() # use default settings

if run_type == "review": 
   ecs_client.execute_review_run()
elif run_type == "final": 
   ecs_client.execute_final_run()
```

**Note:** The PR only appears to have a lot of changes due to the updated dependencies. The main items for review are:
* [implementation of ECSClient in webapp/utils/aws/ecs.py](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/2/files#diff-b90418e2bed6dc92f2077fe7a7720859e6473682f422be94101d9480d229b04d); 
* [added fixtures](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/2/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128) and supporting [unit tests in tests/test_ecs.py](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/2/files#diff-bf5a3552a056fb13827a75297437dd9970de7092832a17ba7d2ec164be0d5d75)

### How can a reviewer manually see the effects of these changes?

1. Run `make test` and verify all tests are passing.

**Note:** For now, running the preliminary set of unit tests is the best way to confirm that the implemented `webapp.utils.aws.ecs.ECSClient` module is working as expected. 

### Includes new or updated dependencies?
YES - Installed libraries: [`attrs`, `boto3-stubs`, `moto`]

### Changes expectations for external applications?
NO

### What are the relevant tickets?
https://mitlibraries.atlassian.net/browse/IN-1015

### Developer
- [X] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -19,6 +30,88 @@ def config():
...
+    with mock_aws():
+        ecs = boto3.client("ecs", region_name=AWS_DEFAULT_REGION)
+        response = ecs.run_task(
+            cluster="mock-sapinvoices-ecs-test",
+            launchType="FARGATE",
+            networkConfiguration=mock_ecs_network_config,
+            overrides={},
+            taskDefinition="mock-sapinvoices-test:1",
+        )
+        yield response["tasks"][0]["taskArn"]
```
Comment (jonavellecuerdo) [inline tests/conftest.py:106]: More specifically, I referenced [this `moto` test example](https://github.com/getmoto/moto/blob/master/tests/test_ecs/test_ecs_boto3.py#L1983-L2045).

File: tests/conftest.py [hunk 2]
```diff
@@ -19,6 +30,88 @@ def config():
...
+
+
+@pytest.fixture
+def ecs_run_task_response_success():
+    with open("tests/fixtures/ecs_run_task_response_success.json") as file:
+        return json.loads(file.read())
```
Comment (jonavellecuerdo) [inline tests/conftest.py:112]: I initially tried to use the `boto3.ECS.client.run_task` mocked output from `moto` but a lot of the response parameters were excluded. I opted to create a fixture by creating a JSON file of the de-identified output from a test run.

File: webapp/utils/aws/ecs.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                    "containerOverrides": [
+                        {
+                            "name": self.container_name,
+                            "command": ["--blahblah"],  # [TODO] Update for SAP Invoices
+                        }
+                    ]
+                },
+                taskDefinition=self.task_definition,
+            )
+            return self.parse_run_details(response)  # type: ignore[arg-type]
```
Comment (jonavellecuerdo) [inline webapp/utils/aws/ecs.py:54]: For now, I've only written a `execute_review_run` method. Later, an `execute_final_run` method will be implemented as well. I am hoping that we can sort out some orchestration + output discussions from a review run first!

File: webapp/utils/aws/ecs.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            task_status = task_run_details["status"]
+
+            message = f"Status for task {task}: {task_status}"
+            logger.info(message)
+
+            if task_status == "STOPPED":
+                logger.info("Task run has completed.")
+                break
+            time.sleep(5)
+        return task_run_details
```
Comment (jonavellecuerdo) [inline webapp/utils/aws/ecs.py:89]: This function will likely be updated once we have a clearer idea of how to display running tasks on the front end (i.e., loading / status updates). For now, this function demonstrates how `boto3.ECS.Client.describe_task` can be used to check on the status of a running ECS task.

File: webapp/config.py [hunk 5]
```diff
@@ -30,6 +31,14 @@ def __getattr__(self, name: str) -> Any:
         message = f"'{name}' not a valid configuration variable"
         raise AttributeError(message)
 
+    def _getenv(self, name: str) -> Any:
+        if value := os.getenv(name):
+            return value
+        if name in self.REQUIRED_ENV_VARS:
+            message = f"'{name}' is a required environment variable."
+            raise AttributeError(message)
+        return None
```
Comment (jonavellecuerdo) [inline webapp/config.py:40]: This is used by the `@property` env access methods. This differs from `__getattr__`, which is only called for attributes that are not explicitly set for the class.

The purpose of `_getenv()` is to implement some error handling so that the `@property` env access methods can have a return type of `str`, which circumvents the following `mypy` error when [`ECSClient` is instantiated](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/2/files#diff-b90418e2bed6dc92f2077fe7a7720859e6473682f422be94101d9480d229b04dR22-R29)

File: webapp/exceptions.py [hunk 6]
```diff
@@ -1,2 +1,9 @@
 class InvalidNetworkConfigurationError(Exception):
     """Exception to raise when ALMA_SAP_INVOICES_ECS_NETWORK_CONFIG is invalid."""
+
+
+class ECSTaskDefinitionDoesNotExistError(Exception):
```
Comment (ehanson8) [inline webapp/exceptions.py:5]: Great custom exception!

File: webapp/utils/aws/ecs.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+from webapp.exceptions import ECSTaskDefinitionDoesNotExistError
+
+logger = logging.getLogger(__name__)
+
+
+@define
+class ECSClient:
+    """ECS Client for running and monitoring task runs."""
+
+    cluster: str = field(factory=lambda: Config().ALMA_SAP_INVOICES_ECS_CLUSTER)
```
Comment (ehanson8) [inline webapp/utils/aws/ecs.py:22]: Just curious since I haven't used it before, what does the `field(factory=lambda:` pattern do here? Is this related to the `@property` decorator?

**Review by ehanson8 — COMMENTED** [review]
A few questions but this is nice and clean!

File: tests/conftest.py [hunk 8]
```diff
@@ -29,6 +49,84 @@ def sapinvoices_client(sapinvoices_app):
...
+
+
+@pytest.fixture
+def mock_ecs_task_state_transitions(
+    mock_ecs_cluster, mock_ecs_task_definition, mock_ecs_network_config
+):
+    """See https://github.com/getmoto/moto/blob/master/tests/test_ecs/test_ecs_boto3.py."""
+    state_manager.set_transition(
+        model_name="ecs::task", transition={"progression": "manual", "times": 1}
+    )
```
Comment (ghukill) [inline tests/conftest.py:110]: This is really cool!  I wasn't aware of a state manager that can cycle through known states for an AWS asset.  I'm going to dig into this a bit more in my own time, become familiar with it, but nice find.  Seems like a perfect use of it.

**Review by ghukill — CHANGES_REQUESTED** [review]
Overall, think this is looking real good, and the logic and structure make sense to me.

Requested some changes (e.g. timeout for `while` loop in monitoring), but most other comments are discussion or optional.

Given that this ECS client is not wired into the app now, I'm feeling good about the structure and functionality at this point.  Maybe when all the pieces come together, there will be some slight tweaks, but this seems like a good base.

File: tests/conftest.py [hunk 9]
```diff
@@ -29,6 +49,84 @@ def sapinvoices_client(sapinvoices_app):
...
+        yield mock_client.return_value
+
+
+@pytest.fixture
+def mock_ecs_cluster():
+    """See https://github.com/getmoto/moto/blob/master/tests/test_ecs/test_ecs_boto3.py."""
+    with mock_aws():
+        ecs = boto3.client("ecs", region_name=AWS_DEFAULT_REGION)
+        ecs.create_cluster(clusterName="mock-sapinvoices-ecs-test")
+        yield
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/conftest.py:69]: I agree! The test examples provided by `moto` are a great resource. ✨

File: tests/conftest.py [hunk 10]
```diff
@@ -29,6 +49,84 @@ def sapinvoices_client(sapinvoices_app):
...
+
+
+@pytest.fixture
+def mock_ecs_task_definition():
+    """See https://github.com/getmoto/moto/blob/master/tests/test_ecs/test_ecs_boto3.py."""
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/conftest.py:74]: Hmm, you are right. I can remove the docstrings. Do you (or @ghukill ) have suggestions for where to share these learnings / resources or maybe it's that the PR itself serves as the documentation for this learning? 😅

File: tests/conftest.py [hunk 11]
```diff
@@ -29,6 +49,84 @@ def sapinvoices_client(sapinvoices_app):
...
+            "subnets": ["subnet-abc123"],
+        }
+    }
+
+
+@pytest.fixture
+def mock_ecs_task_state_transitions(
+    mock_ecs_cluster, mock_ecs_task_definition, mock_ecs_network_config
+):
+    """See https://github.com/getmoto/moto/blob/master/tests/test_ecs/test_ecs_boto3.py."""
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/conftest.py:107]: Docstring will be removed.

File: tests/test_aws/test_ecs.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    assert ecs_client.network_configuration == {
+        "awsvpcConfiguration": {
+            "securityGroups": ["sg-abc123"],
+            "subnets": ["subnet-abc123"],
+        }
+    }
+    assert ecs_client.container == "mock-sapinvoices-test"
+
+
+def test_ecs_client_task_family_access_success(ecs_client):
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/test_aws/test_ecs.py:28]: I like this idea! I renamed the test to `test_ecs_client_task_family_property_success()`.

File: tests/test_aws/test_ecs.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def test_ecs_client_execute_final_run_success(
+    ecs_client, ecs_run_task_response_success, mock_boto3_ecs_client, caplog
+):
+    mock_boto3_ecs_client.run_task.return_value = ecs_run_task_response_success
+
+    with patch(
+        "webapp.utils.aws.ECSClient.task_exists"
+    ) as mock_webapp_ecs_client_task_exists:
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/test_aws/test_ecs.py:57]: Good catch! I created a new fixture called `ecs_client_execute_run_details_success`.

File: webapp/config.py [hunk 14]
```diff
@@ -84,7 +97,7 @@ def configure_logger(self, *, verbose: bool) -> str:
         )
 
     def configure_sentry(self) -> None:
-        if sentry_dsn := self.SENTRY_DSN:
+        if sentry_dsn := self.SENTRY_DSN and self.SENTRY_DSN:
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/config.py:100]: Ah! I meant to remove this update. This will be removed.

File: tests/test_aws/test_ecs.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        ecs_client.task_definition
+        == "arn:aws:ecs:us-east-1:123456789012:task-definition/mock-sapinvoices-ecs-test:1"
+    )
+    assert ecs_client.network_configuration == {
+        "awsvpcConfiguration": {
+            "securityGroups": ["sg-abc123"],
+            "subnets": ["subnet-abc123"],
+        }
+    }
+    assert ecs_client.container == "mock-sapinvoices-test"
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/test_aws/test_ecs.py:25]: Good thought! Updated.

File: webapp/utils/aws/ecs.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    def execute_final_run(self) -> dict | None:
+        logger.info("Executing ECS task for 'Final Run'")
+        return self.run(run_type="final", commands=["--real", "--final"])
+
+    def run(
+        self, run_type: Literal["review", "final"], commands: list | None = None
+    ) -> dict | None:
+        if run_type not in ["review", "final"]:
+            message = f"Cannot run task for unrecognized run_type='{run_type}'"
+            raise RuntimeError(message)
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/utils/aws/ecs.py:53]: Updated.

File: webapp/utils/aws/ecs.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+            time.sleep(5)
+        return task_run_details
+
+    def task_exists(self) -> bool | None:
+        response = self.client.list_task_definitions(
+            familyPrefix=self.task_family, sort="DESC"
+        )
+        valid_tasks = [task.split("/")[-1] for task in response["taskDefinitionArns"]]
+
+        task_family_revision = self.task_definition.split("/")[-1]
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/utils/aws/ecs.py:100]: I added another `@property`-decorated method called `task_family_revision`.

File: tests/conftest.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ghukill) [reply tests/conftest.py]: I'm waffly on this.  I do like linking to documentation sometimes (e.g. linking to Aardvark field definitions in GeoHarvester for quick reference).  But this does feel more inspiration vs. documentation.  I'm fine with them in docstrings, or just PRs with links that identify them as inspiration for tests.

File: tests/conftest.py [hunk 18]
```diff
@@ -29,6 +49,84 @@ def sapinvoices_client(sapinvoices_app):
...
 
+@pytest.fixture
+def ecs_client(mock_ecs_cluster, mock_ecs_task_definition, mock_ecs_network_config):
+    return ECSClient()
+
+
+@pytest.fixture
+def mock_boto3_ecs_client():
+    with patch("webapp.utils.aws.ecs.boto3.client") as mock_client:
+        yield mock_client.return_value
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/conftest.py:60]: Truthfully, I just took this pattern from a fixture I saw in [geo-harvester](https://github.com/MITLibraries/geo-harvester/blob/main/tests/conftest.py#L140-L143). 

To test this theory, I did the following: 
1. Create `webapp/utils/aws/also_ecs.py`.
2. Create an `AlsoECSClient` class:
   ```python
   @define
   class AlsoECSClient:
       cluster: str = field(factory=lambda: Config().ALMA_SAP_INVOICES_ECS_CLUSTER)
       task_definition: str = field(
        factory=lambda: Config().ALMA_SAP_INVOICES_ECS_TASK_DEFINITION
       )
       network_configuration: dict = field(
           factory=lambda: Config().ALMA_SAP_INVOICES_ECS_NETWORK_CONFIG
       )
       container: str = field(factory=lambda: Config().ALMA_SAP_INVOICES_ECR_IMAGE_NAME)

       @property
       def client(self) -> "ECSClientType":
           return boto3.client("ecs")

       def run(self):
           return self.client.run_task(
               cluster=self.cluster,
               launchType="FARGATE",
               networkConfiguration=self.network_configuration,
               overrides={},
               taskDefinition=self.task_definition,
           )
      ```
3. Create a simple test: 
   ```python
   def test_also_ecs_client(mock_boto3_ecs_client):
       also_ecs_client = AlsoECSClient()
       also_ecs_client.run()
   ```

The test runs without error, indicating that the `boto3` client assigned to `AlsoECSClient.client` was still patched.

---

I made the following changes:
1. Renamed fixture `mock_boto3_ecs_client` -> `mock_boto3_client` to avoid confusion on what this fixture does (i.e., patches any `boto3.client` commands--not specific to any service). 
2. Renamed fixture `ecs_client_run_task_response_success` ->  `boto3_ecs_client_run_task_response_success` to clarify that the fixture emulates the response from `boto3.client.ECS.run_task` (not `webapp.utils.aws.ECSClient`). 

Could look into using stubbers instead but for the time being, if these changes are sufficient, I'd like to table making further updates on this. Let me know what you think!

File: webapp/utils/aws/ecs.py [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+        return None
+
+    def monitor_task(self, task: str) -> dict:
+        """Polls ECS for task status updates.
+
+        This method will periodically (every 5 seconds) check for the status
+        of the running ECS task that matches the provided task ARN.
+        When the task run is complete, details of the run are returned.
+        """
+        while True:
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/utils/aws/ecs.py:87]: Let me know what you think of [these updates](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/2/files#diff-b90418e2bed6dc92f2077fe7a7720859e6473682f422be94101d9480d229b04dR86-R89). I implemented a default timeout of 600 seconds (5 minutes). The default is based on the runtime of the most recent [Alma SAP Invoice Processing in Prod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/alma-integrations-sapinvoices-ecs-prod/log-events/sapinvoices$252Falma-integrations-sapinvoices-ecs-prod$252F4d79471f883a47679563ee39942d921f)

[the remark this answers is not in the document]
Comment (ghukill) [reply webapp/utils/aws/ecs.py:87]: I think it looks great!  Particularly like the explicit exception; will be easy to pick out in cloudwatch/sentry if needed.  Thanks for adding.

File: tests/conftest.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ghukill) [reply tests/conftest.py]: Oh embarassing!  My memory is worthless.  Looks familiar... now.  

Re #1: good call (and probably could update in GeoHarvester).  It's definitely mocking the client, which is kind of blank until initialized with a particular client type.

Re #2: Looks good!

Yep, I'm 100% good with tests as-is.

File: webapp/utils/aws/ecs.py (hunk 7 above)
Comment (jonavellecuerdo) [reply webapp/utils/aws/ecs.py:25]: Spent some time reading [documentation on `attrs.Factory` ](https://www.attrs.org/en/stable/api.html#attrs.Factory) but I think @ghukill described it best above in regards to "why `field(default=Config().<ENV>` doesn't work for our unit tests":

> whenever the class (`webapp.config.Config` is first imported / whatever the env vars are at that moment will be the class defaults for those properties forever after, regardless of monkeypatching)

I prefer the current approach over using `__attrs_post_init__` (as shown in @ghukill 's example above); the latter, as the doc describes, 

> **runs** after attrs is done initializing your instance. This is useful if you want to **derive some attribute from others** or perform some kind of **validation over the whole instance**.

I interpret retrieval of the environment 
variables as _part of_ / _the_ initialization step. Conceptually, the use of `__attrs_post_init__` might be confusing. 🤔 

Let me know your final thoughts, @ghukill and @ehanson8 !

File: webapp/utils/aws/ecs.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (ehanson8) [reply webapp/utils/aws/ecs.py:87]: Yes, this is a good approach!

File: webapp/utils/aws/ecs.py (hunk 7 above)
Comment (ehanson8) [reply webapp/utils/aws/ecs.py:25]: @jonavellecuerdo I agree that it is a part of initialization

File: webapp/utils/aws/ecs.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ehanson8) [reply webapp/utils/aws/ecs.py]: I also agree `ValueError` is more appropriate here

**Review by ehanson8 — APPROVED** [review]
Looks good but I would remove `[wip]` from the commit label before merging

File: tests/test_aws/test_ecs.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+    ecs_client, mock_ecs_task_state_transitions, caplog
+):
+
+    task_arn = mock_ecs_task_state_transitions
+
+    with pytest.raises(
+        ECSTaskRuntimeExceededTimeoutError,
+        match="Task runtime exceeded set timeout of 2 seconds.",
+    ):
+        ecs_client.monitor_task(task_arn, timeout=2)
```
Comment (ghukill) [inline tests/test_aws/test_ecs.py:104]: Nice test!  With your 5 second loop, this is a really nice way to exercise this timeout logic.

**Review by ghukill — APPROVED** [review]
Looks great.  Thanks for the updates and discussions.  Looking forward to seeing this wired up in the flask app.
